src/backend/distributed/commands/index.c

```diff
@@ -168,6 +169,93 @@ PlanIndexStmt(IndexStmt *createIndexStatement, const char *createIndexCommand)
 }
 
 
+/*
+ * PlanReindexStmt determines whether a given REINDEX statement involves
+ * a distributed table. If so (and if the statement does not use unsupported
+ * options), it modifies the input statement to ensure proper execution against
+ * the master node table and creates a DDLJob to encapsulate information needed
+ * during the worker node portion of DDL execution before returning that DDLJob
+ * in a List. If no distributed table is involved, this function returns NIL.
+ */
+List *
+PlanReindexStmt(ReindexStmt *ReindexStatement, const char *ReindexCommand)
+{
+	List *ddlJobs = NIL;
+
+	/*
+	 * We first check whether a distributed relation is affected. For that, we need to
+	 * open the relation. To prevent race conditions with later lookups, lock the table,
+	 * and modify the rangevar to include the schema.
+	 */
+	if (ReindexStatement->relation != NULL)
+	{
+		Relation relation = NULL;
+		Oid relationId = InvalidOid;
+		bool isDistributedRelation = false;
+		char *namespaceName = NULL;
+		LOCKMODE lockmode = ShareLock;
```

Comment:
using a ShareLock here might lead to deadlocks, since Postgres itself will take an AccessExclusiveLock, so if we go ShareLock->AccessExclusiveLock we might deadlock with a transaction that goes AccessShareLock->RowExclusiveLock (e.g. SELECT + INSERT).

It probably makes sense to copy the code from ReindexIndex.